src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = {
+          ...currentState,
+          ...action.extraData,
+        };
+        break;
+      case 'removeProperties':
+        currentState = Object.keys(currentState)
+          .filter((key) => !action.keysToRemove.includes(key))
+          .reduce((newState, key) => {
+            newState[key] = currentState[key];
+
+            return newState;
+          }, {});
+        break;
+      default:
+        break;
+    }
+    states.push({ ...currentState });
```

Comment:
The `states.push({ ...currentState });` operation should be moved outside of the `switch` block. According to the additional prompt rules, it is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed.